cpp/src/io/json/process_tokens.cu

```diff
@@ -0,0 +1,313 @@
+              } else if ('.' == c) {
+                num_state = number_state::saw_radix;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::saw_radix:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::fraction;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::fraction:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::fraction;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::start_exponent:
+              if ('+' == c || '-' == c) {
+                num_state = number_state::after_sign_exponent;
+              } else if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::after_sign_exponent:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::exponent:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+          }
+        }
+        return num_state != number_state::after_sign_exponent &&
+               num_state != number_state::start_exponent && num_state != number_state::saw_neg &&
+               num_state != number_state::saw_radix;
+      } else {
+        return false;
+      }
+    };
+
+    auto validate_strings =
+      [data = d_input.data(),
+       allow_unquoted_control_chars =
+         options.is_allowed_unquoted_control_chars()] __device__(SymbolOffsetT start,
+                                                                 SymbolOffsetT end) -> bool {
+      // This validates a quoted string. A string must match https://www.json.org/json-en.html
+      // but we already know that it has a starting and ending " and all white space has been
+      // stripped out. Also the base CUDF validation makes sure escaped chars are correct
+      // so we only need to worry about unquoted control chars
+
+      auto state   = string_state::normal;
+      auto u_count = 0;
+      for (SymbolOffsetT idx = start + 1; idx < end; idx++) {
+        auto c = data[idx];
+        if (!allow_unquoted_control_chars && c < 32) { return false; }
```

Comment:
I have been seeing validation errors on strings with '\x00A0'(Non-breaking_space) in them. I think issue here is two fold, and they kind of overlap with each other.

First C++ does not define if `char` is signed or unsigned. If it ends up being signed, then any value > 127 would be considered invalid because we are not excluding negative values from this check. With how UTF-8 is encoded, this means any character that is not basic ASCII (i.e `\x00A0`). Fixing this alone should be enough to unblock the Spark use cases like from_json, and get_json_object. But it is probably not enough for reading a JSON file directly from python. This is because even with this I am assuming that the supported encodings are ASCII compatible. Can we assume that the input is always UTF-8? Are there any stages in the processing that detect the character encoding and adjust?

Reply:
I don't see NBSP handled anywhere specifically.  To unblock, we can check specifically for NBSP here.
char whether unsigned or signed is implementation specified. If we need a proper comparison, we can cast to unsigned or signed char.
The input could be plain ASCII or UTF-8 (Ascii compatible). The character encoding is not detected anywhere. If there is any parsing error during UTF-8 decoding at strings column creation in json reader, that string value is marked as null.

Reply:
@revans2 should NBSP be handled during white space normalization? is NBSP considered as white space?

Reply:
The whitespace normalization FST only considers `' '`(0x020) and `'\t'`(0x009) as whitespace characters.

Reply:
That is correct the JSON spec (https://www.json.org/json-en.html) defines whitespace as space, linefeed, carriage return, or horizontal tab.

Reply:
@karthikeyann I don't think you fully understand here.

Non blocking space is "\x00A0" in Unicode, but in UTF-8 it is a two byte sequence. -62 and -96.  Both of these end up looking like negative values if char is signed. If it is unsigned then the check is correct.

Reply:
Fixed the issue. 
Comment here for future reference:
`char` whether it is `unsigned`/`signed` is implementation specific as per standard 
> The signedness of char depends on the compiler and the target platform: the defaults for ARM and PowerPC are typically unsigned, the defaults for x86 and x64 are typically signed.

https://en.cppreference.com/w/cpp/language/types#Character_types

So, to write proper comparison expression here, we used `char` type cast here.
```
        if (!allow_unquoted_control_chars && c >= char{0} && c < char{32}) { return false; }
```